### PR TITLE
Add internationalization and localization support

### LINKED ISSUE #3: Configurable Messages File
It could be nice to have a YML file dedicated to configuring the messages displayed by the plugin for various actions, such as initiating a warp, successfully warping, canceling a warp, or getting info of a waystone. This could also be expanded to include the lore on linked Warp Keys.

These would just be simple string values within the file, utilizing some placeholders available through Waystones depending on context, such placeholders could be
```
%waystones_outofrange% - How far out of range a player is from Waystone
%waystones_distance% - Player's total distance from Waystone
%waystones_name% - Name of Waystone in question
%waystones_range% - Range a Waystone is capable of being teleported to from
%waystones_status% - Active status of Waystone, seen in InfoEvent
%waystones_coords% - Coordinates of Waystone, as seen in current WarpKey lore
%waystones_countdown% - Countdown during warmup to Waystone warp
```
These wouldn't have to be global like CMI or PlaceholderAPI, but rather just available through this messages config, the messages would then be parsed to replace these instances when used within the actual events.

This could also allow for custom color codes to be used, though that may require some refactoring of the `Player.sendActionMessage()` function within PlayerUtils.kt

---

The message file could look something like
```yml
InfoMessage: "&bName: %waystones_name% | Status: %waystones_active% | Range: %waystones_range%"
InitWarpMessage: "&aWarping to %waystones_name% in %waystones_countdown% second(s)"
CancelWarpMessage: "&cWarp cancelled"
OutOfRangeMessage: "&c%waystones_name% is out of range (%waystones_outofrange% block(s))"
# etc. etc.
```

---

Slightly unrelated, but this could also open up to the idea of localization, with locale files in a directory holding the messages in different languages. Though I'm unsure as to how this is actually implemented.

### PR DESCRIPTION
**Pull request by NewbieOrange**
Make all messages / actionbars / item displayname and lore customizable.

Some minor modifications were done to internal data classes to facilitate this.

This PR will close #3.

### CODE REVIEW

**Comment by NewbieOrange** [comment]
Hi @AtriusX, any thoughts?

**Comment by AtriusX** [comment]
@NewbieOrange Thanks for the interest in the project! I'll take a look over this one shortly, going to get to the smaller PRs you sent first.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    private var configFile = File(plugin.dataFolder, "locale-$localeTag.yml")
+    private val config = YamlConfiguration()
+    private var locale: Locale = Locale.getDefault()
+
+    init {
+        reload(localeTag)
+    }
+
+    fun localize(key: String, vararg args: Any?): String {
+        return MessageFormat(config.getString(key)!!, locale)
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:21]: You should avoid using the `!!` operator since it can cause exceptions to be thrown if the key references a nonexistent value. In this case you should consider passing a default value using the `?:` operator.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                .format(*args.map { it ?: "null" }.toTypedArray()).translateColors()
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:22]: This feels somewhat inflexible to me. There is at least one case I can think of where we would want the value to be printed with a different value. Additionally, "null" wouldn't get localized into the appropriate languages.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    fun reload(localeTag: String) {
+        configFile = File(plugin.dataFolder, "locale-$localeTag.yml")
+
+        if (!configFile.exists()) {
+            plugin.saveResource(configFile.name, false)
+            config.load(configFile)
+        } else {
+            val defaultConfigFile = plugin.getResource(configFile.name) ?: plugin.getResource("locale-en.yml")!!
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:32]: Please avoid use of the `!!` operator as mentioned above.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            val defaultConfig = YamlConfiguration.loadConfiguration(InputStreamReader(defaultConfigFile))
+            config.load(configFile)
+            config.options().copyDefaults(true)
+            config.setDefaults(defaultConfig)
+            config.save(configFile)
+        }
+
+        locale = Locale.forLanguageTag(config.getString("locale"))
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:40]: As mentioned earlier I think it might make more sense to delegate this to the configuration using a `LocaleParser` object.

File: src/main/kotlin/xyz/atrius/waystones/handler/WaystoneHandler.kt [hunk 5]
```diff
@@ -79,5 +79,5 @@ class WaystoneHandler(
     }
 
     private fun distanceError(name: String, distance: Double, range: Int): String =
-        "$name is out of warp range [%.1f block(s)]".format(distance - range)
-}
\ No newline at end of file
+        localization.localize("warp-out-of-range", name, distance - range)
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/handler/WaystoneHandler.kt:82]: A bit concerned about this line. Does this preserve the single-decimal rounding from the original code?

**Review by AtriusX — CHANGES_REQUESTED** [review]
First pass on the PR. I also noticed that you were calling `localization.localize()` a lot. Given the nature of the object, it might be a bit better to implement an operator function on `get()` so that you can reference localizations like this: `localization["localized-string", 1, 2, "test"]`
I feel like this might help cut down on the verbosity a bit and it would bring it a bit closer to matching the way we handle config properties.

File: src/main/kotlin/xyz/atrius/waystones/data/KeyState.kt [hunk 6]
```diff
@@ -1,14 +1,15 @@
...
 
 sealed class KeyState(val message: String?) {
 
     object None : KeyState(null)
 
-    object Severed : KeyState("The link to this waystone has been severed")
+    object Severed : KeyState(localization.localize("key-severed"))
 
-    object Blocked : KeyState("You are too sick to warp")
+    object Blocked : KeyState(localization.localize("key-blocked"))
```
[the remark this answers is not in the document]
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/KeyState.kt:12]: If needed, please avoid marking these as `var`. It would likely make more sense to convert these to a high-order function. Example:

```kt
sealed class KeyState(val message: () -> String?)

// ...

/*
 * If the last argument is a high-order function you should be able to place it outside 
 * of the (), and omit them entirely if the function is the only argument passed. I'm not
 * completely sure how this behaves with constructors specifically though.
 */
object Severed : KeyState { localization.localize("key-severed") }
```

File: src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt [hunk 7]
```diff
@@ -1,43 +1,41 @@
...
+            localization.localize("waystone-status-obstructed"))
 
     open fun message(name: String): String =
         message.format(name)
 }
 
 sealed class WarpActiveState(
     val range: Int
-) : WarpState("Active") {
+) : WarpState(localization.localize("waystone-status-active"), range) {
```
[the remark this answers is not in the document]
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt:36]: Yes, check the locale files.

File: src/main/kotlin/xyz/atrius/waystones/handler/WaystoneHandler.kt (hunk 5 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/handler/WaystoneHandler.kt:82]: Yes.

File: src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt [hunk 8]
```diff
@@ -1,43 +1,41 @@
...
+        return localization.localize("waystone-status", status, range)
     }
 }
 
 sealed class WarpErrorState(
     private val message: String,
-    status: String = "Unknown",
-) : WarpState(status) {
+    status: String = localization.localize("waystone-status-unknown"),
+) : WarpState(status, -2) {
```
[the remark this answers is not in the document]
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt:17]: The base.toString() needs to pass range into the string template.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt [hunk 9]
```diff
@@ -13,6 +13,10 @@ class Config(plugin: KotlinPlugin) {
         plugin.saveDefaultConfig()
     }
 
+    // The localization to be used
+    val localization: Property<String> =
+        Property("localization", "en", StringParser)
+
```
[the remark this answers is not in the document]
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt:19]: Here a string is used in case the user want to create their own language file and the filename is not any `Locale`.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 1 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:21]: The `config.getString(key)` call should never return null since we applied in-jar defaults to the config. If it is null then the plugin is trying to read a non-existent string, which imo better not to fail silently.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 2 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:22]: True. The infleibility can be somewhat mitigated tho, check the locale-en.yml. For `null`, maybe avoid passing null directly into the localization class? Instead ?: into a proper representation based on the context (e.g. unnamed-waystone: "Waystone").

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 3 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:32]: If there is no `locale-en.yml` in the jar, an Exception should probably be thrown?

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 1 above)
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:21]: I can understand the rationale, but looking at it as an error using the `!!` a bit vague since it just returns a `NullPointerException`. Perhaps instead an error should be explicitly thrown instead? Example: ```kt
```kt
return MessageFormat(config.getString(key) ?: throw IllegalArgumentException("$key does not exist in configuration file!"), locale)
```

File: src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt:17]: So this is used for the logic in the locale files then, alright cool.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt:19]: I believe java's `Locale` class should support the creation of custom locale types. The primary thing I think I'm concerned about is following some level of standardization with the file names.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 2 above)
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:22]: A fair point, I'd recommend going through things again to see if there is any place that might be occurring that could've been missed.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 3 above)
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:32]: If we should throw an exception I think it would be best to ensure its clear in its meaning by throwing our own instead of relying on a `NullPointerException`.

File: src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/WarpState.kt:36]: Alright

File: src/main/kotlin/xyz/atrius/waystones/data/KeyState.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/KeyState.kt:12]: I was thinking about returning a `LocalizedString` instead, which overrides invoke() and toString() to return `localization.localize` result. It may be better than returing function references.

[the remark this answers is not in the document]
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/KeyState.kt:12]: I'm not sure honestly, you'd need to still call toString() implicitly in order for it to function as expected anyways. Depending how you write it, it could be a fairly hacky solution too.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt [hunk 10]
```diff
@@ -11,15 +11,24 @@ import java.util.*
...
+                ?: throw IllegalArgumentException("$key does not exist in localization file!")
+        return format(template, *args)
+    }
+
+    fun format(string: String, vararg args: Any?): String {
+        return MessageFormat(string, locale).format(args).translateColors()
+    }
+
+    operator fun get(key: String, vararg args: Any?): String {
+        return localize(key, *args)
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt]: You mentioned before that we should avoid passing `null` into the formatting methods, it might be a good idea to enforce that at compile-time as well by switching out `Any?` with just `Any`

File: src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt:19]: I have added `LocalizedString` class, which seems to work fine.

File: src/main/kotlin/xyz/atrius/waystones/data/KeyState.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/KeyState.kt]: I have added LocalizedString class, which seems to work fine.

File: src/main/kotlin/xyz/atrius/waystones/handler/LinkHandler.kt [hunk 11]
```diff
@@ -23,14 +23,14 @@ class LinkHandler(
...
+            return Fail(localization["link-no-permission"].toString())
         // Prevent linking if relinking is disabled
         val meta = item.itemMeta as CompassMeta
         return if (!configuration.relinkableKeys() && meta.hasLodestone())
-            Fail("The destination for this key has been sealed")
+            Fail(localization["link-not-relinkable"].toString())
         // Prevent relinking if the location is the same
         else if (!player.immortal && meta.lodestone == block.location)
-            Fail("Already linked to this destination")
+            Fail(localization["link-already-linked"].toString())
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/handler/LinkHandler.kt:33]: We seem to be calling `toString()` a lot; i'm not sure how much I like this approach. We might need to find a better way to handle this or possibly convert any message properties to `LocalizedString`.

**Review by AtriusX — COMMENTED** [review]
Checked over the changes

File: src/main/kotlin/xyz/atrius/waystones/handler/LinkHandler.kt (hunk 11 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/handler/LinkHandler.kt]: Now the Fail() takes a `LocalizedString` instead, no more `toString()` other than in utility classes.

File: gradle/wrapper/gradle-wrapper.properties [hunk 12]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.6.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.5.1-bin.zip
```
Comment (AtriusX) [inline gradle/wrapper/gradle-wrapper.properties:3]: What is the reason for this change?

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 10 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt]: There are a lot of varaible could be null (while actually never is) in current code, changing them seems not in the scope of this PR.

File: gradle/wrapper/gradle-wrapper.properties (hunk 12 above)
Comment (NewbieOrange) [reply gradle/wrapper/gradle-wrapper.properties:3]: Gradle 6.6/6.7 cause errors with kotlin on some specific java versions: gradle/gradle#14155.

Either 6.5.1 or 6.8.3 works for me.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 4 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:59]: The issue is YamlConfig will save the Locale as a serialized java object, which is unfriendly to users.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 10 above)
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt]: That's fair

File: gradle/wrapper/gradle-wrapper.properties (hunk 12 above)
Comment (AtriusX) [reply gradle/wrapper/gradle-wrapper.properties:3]: I see. It might be better to stick to the more recent version if that is the case then.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt (hunk 4 above)
Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:59]: The parser interface has a `toString(value: T)` method you can override which can be implemented for rendering the object to a valid string. The parser system should support two-way conversions as a result.

Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:59]: I see, missed that part.

**Comment by AtriusX** [comment]
Added a comment to commit 0abd193

**Comment by NewbieOrange** [comment]
Does commit 8118eae break expected behaviour?

File: src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt [hunk 13]
```diff
@@ -10,14 +10,13 @@ import java.text.MessageFormat
 import java.util.*
 import kotlin.collections.HashMap
 
-class Localization(val plugin: KotlinPlugin, localeTag: String) {
-    private var configFile = File(plugin.dataFolder, "locale-$localeTag.yml")
+class Localization(private val plugin: KotlinPlugin, private val locale: Locale) {
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Localization.kt:13]: We should probably be referencing this directly from the configuration property to ensure we are pulling the most recent locale. Reloading should probably reference the config property as well.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt [hunk 14]
```diff
@@ -59,6 +60,12 @@ object BooleanParser : ArgumentParser<Boolean> {
         if (input?.toLowerCase() in arrayOf("true", "false")) input.toBoolean() else null
 }
 
+object LocaleParser : ArgumentParser<Locale> {
+    override fun parse(input: String?): Locale? = input?.let { Locale.forLanguageTag(input) }
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt:64]: We should probably enforce some form of format before setting the value as mentioned in another comment

**Comment by AtriusX** [comment]
> 
> 
> Does commit [8118eae](https://github.com/AtriusX/Waystones/commit/8118eae50b5ff90e9292300e628edf0817545fed) break expected behaviour?

I can't really confirm but I don't see why it would break. Should test it just to be safe though.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt (hunk 14 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt:64]: `Locale.forLanguageTag()` accepts any `String`, there will be no error thrown for sure.

If the Locale is something that `MessageFormat` does not recognize, it will probably treat it in a generic way.

Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt:64]: Any suggestions for what we should this?

**Comment by AtriusX** [comment]
Will be going offline shortly, so any changes afterward will be looked at in the morning. 👋🏼

File: src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt (hunk 14 above)
Comment (NewbieOrange) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt:64]: Some documentation about this behaviour should be enough.

Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Parsers.kt:64]: Probably should add documentation for it to the README file then.

**Comment by AtriusX** [comment]
We still have to add an update to the README file, but otherwise we can try testing things as they are. @Sepshun mind giving this a look?

**Comment by Sepshun** [comment]
This is taking a while as I'm trying to rather extensively test it, since this is a decent sized system being implemented.
So, a few points:

I'm noticing one small inconsistency though, not sure if this is intentional or not but upon changing the localization value through `/ws config localization [language]`, it doesn't reload and take affect immediately, though every other config value type does.
You could just run the reload command to and that'd take care of it, but afaik it does this for every value other than localization.

Potentially unrelated to this PR but rather the way we have the WarpState system setup, but no error is being displayed at all on a warp that has been severed/broken.

Other than this, everything else functions incredibly well!

**Comment by AtriusX** [comment]
> I'm noticing one small inconsistency though, not sure if this is intentional or not but upon changing the localization value through `/ws config localization [language]`, it doesn't reload and take effect immediately, though every other config value type does.
> You could just run the reload command to and that'd take care of it, but afaik it does this for every value other than localization.

Maybe this has to do with the localization class not being stored within the config itself? A solution to this might be to retrieve the current locale within the `get` operator of the class and reload if it is different. Otherwise we can just assign the update by moving the storage of the class into the configuration instead of using `Locale`.

**Comment by AtriusX** [comment]
@NewbieOrange Any thoughts on updating `LocaleParser` to store `Localization` objects instead? This would remove the need to manually reload it since the object would be regenerated each time the option is updated.

**Comment by NewbieOrange** [comment]
I am thinking about make Property have something like onUpdate() and we can pass localization.reload() as an argument, which the Property will call on update.

**Comment by AtriusX** [comment]
That could work too actually, try that out if you'd like

**Comment by AtriusX** [comment]
Looking at the code you might just need to apply this to the relevant files? 
```kt
class Property<T>(
    property: String,
    default: T,
    private val parser: ArgumentParser<T>,
    private val onUpdate: () -> Unit = {}
) {
	// ..........
    operator fun invoke(input: String?): Boolean {
        value = parser.parse(input) ?: return false
        onUpdate()
        return true
    }
```
And then apply this to the configuration:
```kt
val localization: Property<Locale> =
        Property("localization", Locale.ENGLISH, LocaleParser) {
			localization.reload()
		}
```
From what I can tell, this might also remove the need to explicitly call `localization.reload()` in the reload command based off how the `ConfigManager` reloads data. (It might actually call `onUpdate` automatically as a result.

**Comment by NewbieOrange** [comment]
I have updated the code as you suggested, was busy working on other stuff.

I don't have time to test whether it work or not yet.

File: src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt [hunk 15]
```diff
@@ -15,8 +16,8 @@ class Config(plugin: KotlinPlugin) {
     }
 
     // The localization to be used
-    val localization: Property<Locale> =
-        Property("localization", Locale.ENGLISH, LocaleParser)
+    val locale: Property<Locale> =
+        Property("localization", Locale.ENGLISH, LocaleParser, { localization.reload() })
```
Comment (AtriusX) [inline src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt]: I noticed you renamed this to `locale`, so you should also update `"localization"` to match the rename as that is the value that determines the property it saves and looks for in the config. Additionally it's preferable to keep the `onUpdate` outside the parenthesis as that is generally how kotlin handles function arguments.

Comment (AtriusX) [reply src/main/kotlin/xyz/atrius/waystones/data/config/Config.kt]: On second look I didn't see the 2nd commit which fixed this.

**Comment by AtriusX** [comment]
That's fine, @Sepshun can test the update for us again. Once we know that this works then we just need an update to the `README` file I think.

**Comment by Sepshun** [comment]
Seems to be working perfectly!

**Comment by AtriusX** [comment]
@NewbieOrange can you add the option to the README file? I think that should finish things off.

**Comment by NewbieOrange** [comment]
Added the option to README. Note that English is not my first language, feel free to update the description.

**Review by AtriusX — APPROVED** [review]
Looks pretty good now!
